### PR TITLE
chore: Adds end-to-end tests for search deployment

### PR DESCRIPTION
**Pull request by lantoli**
## Proposed changes

Adds end-to-end tests for search deployment

_Jira ticket:_ CLOUDP-234049

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] This change requires a documentation update
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

## Manual QA performed:

- [ ] cfn invoke for each of CRUDL/cfn test
- [ ] Updated resource in  [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples)
- [ ] Published to AWS private registry
- [ ] Used the template in [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples) to create and update a stack in AWS
- [ ] Deleted stack to ensure resources are deleted
- [ ] Created multiple resources in same stack
- [ ] Validated in Atlas UI
- [ ] Included screenshots

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run `make fmt` and formatted my code
- [ ] For CFN Resources: I have released by changes in the private registry and proved by change
  works in Atlas

### CODE REVIEW

File: cfn-resources/test/e2e/utility/atlas_helper.go [hunk 1]
```diff
@@ -57,14 +59,85 @@ func NewMongoDBClient() (atlasClient *admin.APIClient, err error) {
...
-	sdkV2Client, err := c.NewSDKV2Client(client)
+	sdkV2Client, err := c.NewSDKV2LatestClient(client)
 	if err != nil {
 		return nil, fmt.Errorf("unable to create Atlas client")
 	}
 
 	return sdkV2Client, nil
 }
 
+func CreateProject(t *testing.T, atlasClient *admin.APIClient, orgID, projectName string) string {
```
Comment (marcosuma) [inline cfn-resources/test/e2e/utility/atlas_helper.go:70]: it is just curious that we didn't have this code implemented in CFN to create resources for tests. Well done @lantoli @AgustinBettati!! - Am I correct to say this is the first e2e test we're writing for CFN?

File: cfn-resources/test/e2e/utility/atlas_helper.go [hunk 2]
```diff
@@ -35,7 +37,7 @@ func NewAtlasTeam(ctx context.Context, client *admin.APIClient, name string, org
 	orgUser, _ := getExistingOrgUser(ctx, client, orgID)
 	teamRequest := admin.Team{
 		Name:      name,
-		Usernames: []string{orgUser.Username},
+		Usernames: &[]string{orgUser.Username},
```
Comment (marcosuma) [inline cfn-resources/test/e2e/utility/atlas_helper.go:40]: why was this needed?

**Review by marcosuma — APPROVED** [review]

File: cfn-resources/test/e2e/search-deployment/searchdeployment_template.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
+{
+  "AWSTemplateFormatVersion": "2010-09-09",
+  "Description": "This template creates a Search Index on the MongoDB Atlas API, this will be billed to your Atlas account.",
```
Comment (AgustinBettati) [inline cfn-resources/test/e2e/search-deployment/searchdeployment_template.json:3]: we can rename this description, can see it was a miss coming from the examples section

File: cfn-resources/test/e2e/utility/atlas_helper.go (hunk 1 above)
Comment (lantoli) [reply cfn-resources/test/e2e/utility/atlas_helper.go:70]: this is the second one, the first one for project resource was created by @maastha 

the idea we're following is to test in CFN the specific resource, and get the prerequisites (like project, cluster) ready  for it to use

File: cfn-resources/test/e2e/search-deployment/searchdeployment_test.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	resourceCtx    utility.ResourceContext
+	template       string
+}
+
+type testProject struct {
+	ResourceTypeName string
+	Profile          string
+	ProjectID        string
+	ClusterName      string
+}
```
Comment (AgustinBettati) [inline cfn-resources/test/e2e/search-deployment/searchdeployment_test.go:40]: Could we add InstanceSize and NodeCount here? Leave as dynamic value in the stack template, and then we can assert the values in the `testCreateStack`

File: cfn-resources/test/e2e/search-deployment/searchdeployment_test.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+	t.Run("Validate Template", func(t *testing.T) {
+		utility.TestIsTemplateValid(t, testCtx.cfnClient, testCtx.template)
+	})
+
+	t.Run("Create Stack", func(t *testing.T) {
+		testCreateStack(t, testCtx)
+	})
+
+	t.Run("Delete Stack", func(t *testing.T) {
```
Comment (AgustinBettati) [inline cfn-resources/test/e2e/search-deployment/searchdeployment_test.go:68]: we can include an update check here, mainly just modifying the InstanceSize or NodeCount

**Review by AgustinBettati — APPROVED** [review]
LGTM, leaving some possible improvements

File: cfn-resources/test/e2e/search-deployment/searchdeployment_template.json (hunk 3 above)
Comment (lantoli) [reply cfn-resources/test/e2e/search-deployment/searchdeployment_template.json:3]: yes, this is from the examples, do you have any suggestion?

File: cfn-resources/test/e2e/search-deployment/searchdeployment_test.go (hunk 4 above)
Comment (lantoli) [reply cfn-resources/test/e2e/search-deployment/searchdeployment_test.go:40]: InstanceSize for sure, i had a problem with NodeCount because it's an int and the JSON was bad formed, let's discuss offline

File: cfn-resources/test/e2e/search-deployment/searchdeployment_test.go (hunk 5 above)
Comment (lantoli) [reply cfn-resources/test/e2e/search-deployment/searchdeployment_test.go:68]: let's discuss offline, i removed it because i couldn't change NodeCount

Comment (lantoli) [reply cfn-resources/test/e2e/search-deployment/searchdeployment_test.go:68]: i will add the update operation, and will change the template so extension is not json

File: cfn-resources/test/e2e/search-deployment/searchdeployment_template.json (hunk 3 above)
Comment (lantoli) [reply cfn-resources/test/e2e/search-deployment/searchdeployment_template.json]: thx, updated

File: cfn-resources/test/e2e/search-deployment/searchdeployment_test.go (hunk 4 above)
Comment (lantoli) [reply cfn-resources/test/e2e/search-deployment/searchdeployment_test.go]: updated

**Comment by maastha** [comment]
LGTM, this is super!
Would only add that we should ensure that cleanup script is invoked for any kind of failures/errors appropriately because otherwise we will be left with dangling resources published to our private registry which have a certain limit as well (I think ~50)

File: cfn-resources/test/e2e/utility/atlas_helper.go (hunk 1 above)
Comment (marcosuma) [reply cfn-resources/test/e2e/utility/atlas_helper.go:70]: Awesome @maastha!

**Review by AgustinBettati — APPROVED** [review]

**Comment by lantoli** [comment]
@maastha while I was developing in local it failed in many places and all resources were deleted so i think we're fine
